src/d3d9/shaders/d3d9_fixed_function_frag.frag

```diff
@@ -0,0 +1,667 @@
+const uint D3DTOP_BLENDTEXTUREALPHAPM       = 15;
+const uint D3DTOP_BLENDCURRENTALPHA         = 16;
+const uint D3DTOP_PREMODULATE               = 17;
+const uint D3DTOP_MODULATEALPHA_ADDCOLOR    = 18;
+const uint D3DTOP_MODULATECOLOR_ADDALPHA    = 19;
+const uint D3DTOP_MODULATEINVALPHA_ADDCOLOR = 20;
+const uint D3DTOP_MODULATEINVCOLOR_ADDALPHA = 21;
+const uint D3DTOP_BUMPENVMAP                = 22;
+const uint D3DTOP_BUMPENVMAPLUMINANCE       = 23;
+const uint D3DTOP_DOTPRODUCT3               = 24;
+const uint D3DTOP_MULTIPLYADD               = 25;
+const uint D3DTOP_LERP                      = 26;
+
+const uint D3DTA_SELECTMASK     = 0x0000000f;
+const uint D3DTA_DIFFUSE        = 0x00000000;
+const uint D3DTA_CURRENT        = 0x00000001;
+const uint D3DTA_TEXTURE        = 0x00000002;
+const uint D3DTA_TFACTOR        = 0x00000003;
+const uint D3DTA_SPECULAR       = 0x00000004;
+const uint D3DTA_TEMP           = 0x00000005;
+const uint D3DTA_CONSTANT       = 0x00000006;
+const uint D3DTA_COMPLEMENT     = 0x00000010;
+const uint D3DTA_ALPHAREPLICATE = 0x00000020;
+
+const uint D3DRTYPE_SURFACE       = 1;
+const uint D3DRTYPE_VOLUME        = 2;
+const uint D3DRTYPE_TEXTURE       = 3;
+const uint D3DRTYPE_VOLUMETEXTURE = 4;
+const uint D3DRTYPE_CUBETEXTURE   = 5;
+const uint D3DRTYPE_VERTEXBUFFER  = 6;
+const uint D3DRTYPE_INDEXBUFFER   = 7;
+
+const uint VK_COMPARE_OP_NEVER            = 0;
+const uint VK_COMPARE_OP_LESS             = 1;
+const uint VK_COMPARE_OP_EQUAL            = 2;
+const uint VK_COMPARE_OP_LESS_OR_EQUAL    = 3;
+const uint VK_COMPARE_OP_GREATER          = 4;
+const uint VK_COMPARE_OP_NOT_EQUAL        = 5;
+const uint VK_COMPARE_OP_GREATER_OR_EQUAL = 6;
+const uint VK_COMPARE_OP_ALWAYS           = 7;
+
+
+// Bindings have to match with computeResourceSlotId in dxso_util.h
+// computeResourceSlotId(
+//     DxsoProgramType::PixelShader,
+//     DxsoBindingType::ConstantBuffer,
+//     DxsoConstantBuffers::PSFixedFunction
+// ) = 11
+layout(set = 0, binding = 11, scalar, row_major) uniform ShaderData {
+    D3D9FixedFunctionPS data;
+};
+
+// Bindings have to match with computeResourceSlotId in dxso_util.h
+// computeResourceSlotId(
+//     DxsoProgramType::PixelShader,
+//     DxsoBindingType::ConstantBuffer,
+//     DxsoConstantBuffers::PSShared
+// ) = 12
+layout(set = 0, binding = 12, scalar, row_major) uniform SharedData {
+    D3D9SharedPS sharedData;
+};
+
+layout(push_constant, scalar, row_major) uniform RenderStates {
+    D3D9RenderStateInfo rs;
+
+    layout(offset = MaxSharedPushDataSize) uint packedSamplerIndices[TextureStageCount / 2];
+};
+
+layout(set = 0, binding = 13) uniform texture2D t2d[TextureStageCount];
+layout(set = 0, binding = 13) uniform textureCube tcube[TextureStageCount];
+layout(set = 0, binding = 13) uniform texture3D t3d[TextureStageCount];
+
+layout(origin_upper_left) in vec4 gl_FragCoord;
+
+layout(set = 15, binding = 0) uniform sampler sampler_heap[];
```

Comment:
I know we do this in the legacy compilers as well, but there's some tooling that shits the bed with `set = 15`, so might be a good idea to just set this to 1 if doesn't mess with the plumbing code too much.